### DISCUSSION TITLE
Inquiry about 'diff-ergmTerm' in ERGM package

### DISCUSSION THREAD

**Discussion by dleeao** [statnet/ergm#525] [General]
Dear,

 I hope this email finds you well. I am currently a student at KAIST, studying network theory using the ERGM package in R language. I wanted to express my gratitude for your packages in my studies.

I have a question regarding the 'diff-ergmTerm' in the ERGM package, specifically related to the directionality of the tail-head and the use of posonly and negonly.

 I created a network with 19 nodes (numbered from 1 to 19) and assigned sequential attributions to each node, with higher attribution values for smaller node numbers (e.g., 1 having the highest attribution and the attribution decreasing as the node number increases).
 Then, I generated edges using code to create links with higher probabilities from nodes with higher attribution to nodes with lower attribution.
 Specifically, I designed the links to have a higher probability of tail(i)-head(j) with a positive sign.action(attr[i]-attr[j]) in the 'diff' function.

 However, when I included both "diff('Att', pow=1, dir="t-h", sign.action="posonly")" and "diff('Att', pow=1, dir="t-h", sign.action="negonly")" in the code and ran it, I obtained unexpected results, where 'negonly' showed statistically significant effects instead of 'posonly'.

I would appreciate it if you could review my code and data set to confirm if my interpretation and implementation are correct. Please find the attached data set and code for your reference.

Thank you in advance for your assistance.

Best regards,
Deok-Young Lee.

Datasets:
[test_edge.csv](https://github.com/statnet/ergm/files/11238937/test_edge.csv)
[test_node.csv](https://github.com/statnet/ergm/files/11238938/test_node.csv)
Result I got:
[diff_results](https://user-images.githubusercontent.com/109405991/232203378-fa4f5b73-8b9d-4c5b-b014-f9fc8bbd6535.png)

Codes:
```R
library('ergm')
library('igraph')
library('intergraph')

link_test <- read.csv("test_edge.csv")
G <- graph.data.frame(link_test[, c(1,2)], directed = TRUE)
property_test <- read.csv("test_node.csv")
colnames(property_test)
V(G)$Att <- property_test[,2]

nx_test <- asNetwork(G, ignore.eval=FALSE)
nx_test

base_test <- ergm(nx_test ~ edges)
base_test_diffsign <- ergm(nx_test ~ edges 
                           + diff('Att', pow=1, dir="t-h", sign.action="posonly")
                           + diff('Att', pow=1, dir="t-h", sign.action="negonly"))
summary(base_test_diffsign)
```

**Comment by mbojan** [comment]
I believe you have perfect separation in your data, i.e. tie exist only in those dyads in which the difference on 'att' is positive.

**Comment by dleeao** [comment]
> I believe you have perfect separation in your data, i.e. tie exist only in those dyads in which the difference on 'att' is positive.

Yes, that’s right. I checked again and found that there is a perfect separation. Nevertheless, in the ergm results, only 'negonly' seems to be statistically significant (see the results in the original article), so I ask a question.
https://cran.r-project.org/web/packages/ergm/ergm.pdf
According to the link above ('diff-ergmTerm'), seeing that it says:
["posonly": positive differences are kept, negative differences are replaced by 0]
, I thought that only 'posonly' should be statistically significant.

**Comment by mbojan** [comment]
The estimates, and the SEs are meaningless in this case because the MPLE does not exist.

**Comment by dleeao** [comment]
Thank you for the good point. I'm still studying ERGM, and I'm a beginner, so I think there are some parts I can't think of. In fact, the data I got is as follows. The diff was carried out using the attributes 'AR' and 'EC'.
```R
Call:
ergm(formula = nx~ sum + sum(pow = 1/2) + mutual + 
    nodeocov("AR", form = "sum") + diff("AR", pow = 1, dir = "t-h", 
    sign.action = "posonly", form = "sum") + diff("AR", pow = 1, 
    dir = "t-h", sign.action = "negonly", form = "sum") + nodeocov("ER", 
    form = "sum") + nodeocov("C", form = "sum") + nodeocov("H", 
    form = "sum") + nodeocov("EC", form = "sum") + diff("EC", 
    pow = 1, dir = "t-h", sign.action = "posonly", form = "sum") + 
    diff("EC", pow = 1, dir = "t-h", sign.action = "negonly", 
        form = "sum"), response = "weight", reference = ~Poisson, 
    control = control.ergm(MCMC.interval = 1024, MCMC.burnin = 5e+06, 
        MCMC.samplesize = 20000, MCMLE.maxit = 30, seed = 2023))

Iterations:  8 out of 30 

Monte Carlo MLE Results:
                        Estimate Std. Error MCMC % z value Pr(>|z|)    
sum                     -2.76503    0.21551      0 -12.830  < 1e-04 ***
sum0.5                  -4.25322    0.05441      0 -78.176  < 1e-04 ***
mutual.min               1.55081    0.05160      0  30.057  < 1e-04 ***
nodeocov.sum.AR          1.79829    0.19005      0   9.462  < 1e-04 ***
diff.sum.posonly.t-h.AR  0.37422    0.10057      0   3.721 0.000199 ***
diff.sum.negonly.t-h.AR  2.83449    0.23907      0  11.857  < 1e-04 ***
nodeocov.sum.ER          0.39356    0.14808      0   2.658 0.007868 ** 
nodeocov.sum.C           2.00138    0.27969      0   7.156  < 1e-04 ***
nodeocov.sum.H           0.68073    0.33104      0   2.056 0.039746 *  
nodeocov.sum.EC          0.18083    0.08860      0   2.041 0.041251 *  
diff.sum.posonly.t-h.EC -0.40171    0.04529      0  -8.869  < 1e-04 ***
diff.sum.negonly.t-h.EC  0.56378    0.11511      0   4.898  < 1e-04 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1
```
In this case, too, dyads in which the difference on 'att' is positive with a higher probability, 
so (I thought) 'posonly' would have positive Estimates.
But, with different results than intuition, I created a test dataset like the original question.

By any chance, please let me know if there is anything I am not understanding.
Thank you.

**Reply by mbojan** [reply]
Notice that the signs of `diff.posonly` and `diff.negonly` essentially suggest that ties are the more likely the closer to 0 is the difference sender-receiver on the attribute, This is a bit like asymmetric `absdiff` term as the slope on the left hand side of 0 is a bit steeper than on the right hand of 0, that asymmetry does not seem too strong though,
